[BUG] operator crashes if tls config is invalid

If tls config is invalid,
the operator just crashes instead of giving a proper validation error.

```markdown
2022-09-09T16:22:38.660Z	DEBUG	controllers/replica_set_controller.go:140	Validating MongoDB.Spec	{"ReplicaSet": "..."}
2022-09-09T16:22:38.661Z	DEBUG	controllers/replica_set_controller.go:149	Ensuring the service exists	{"ReplicaSet": "..."}
2022-09-09T16:22:38.864Z	DEBUG	agent/replica_set_port_manager.go:122	No port change required	{"ReplicaSet": "..."}
2022-09-09T16:22:38.874Z	INFO	controllers/replica_set_controller.go:464	Create/Update operation succeeded	{"ReplicaSet": "...", "operation": "updated"}
2022-09-09T16:22:38.874Z	INFO	controllers/mongodb_tls.go:39	Ensuring TLS is correctly configured	{"ReplicaSet": "..."}
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x13d202e]

goroutine 354 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:118 +0x1f4
panic({0x15247e0, 0x24553c0})
	/usr/local/go/src/runtime/panic.go:838 +0x207
github.com/mongodb/mongodb-kubernetes-operator/api/v1.MongoDBCommunity.TLSConfigMapNamespacedName(...)
	/workspace/api/v1/mongodbcommunity_types.go:797
github.com/mongodb/mongodb-kubernetes-operator/controllers.getCaCrt({_, _}, {_, _}, {{{0x14080c8, 0x10}, {0xc0004b61e0, 0x1f}}, {{0xc0003f9d20, 0x10}, ...}, ...})
	/workspace/controllers/mongodb_tls.go:166 +0x12e
github.com/mongodb/mongodb-kubernetes-operator/controllers.(*ReplicaSetReconciler).validateTLSConfig(_, {{{0x14080c8, 0x10}, {0xc0004b61e0, 0x1f}}, {{0xc0003f9d20, 0x10}, {0x0, 0x0}, {0xc0003f9d18, ...}, ...}, ...})
	/workspace/controllers/mongodb_tls.go:42 +0x125
github.com/mongodb/mongodb-kubernetes-operator/controllers.ReplicaSetReconciler.Reconcile({{0x19c5b60, 0xc0005248a0}, 0xc000158540, 0xc0003fe158, 0xc0003fe370, 0xc0003fe378}, {0x19bd600?, 0xc000130d20?}, {{{0xc0003f9d18, 0x8}, ...}})
	/workspace/controllers/replica_set_controller.go:158 +0x7d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x19bd558?, {0x19bd600?, 0xc000130d20?}, {{{0xc0003f9d18?, 0x1627a00?}, {0xc0003f9d20?, 0x4041f4?}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:121 +0xc8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0004a28c0, {0x19bd558, 0xc000526f00}, {0x1574ca0?, 0xc000348060?})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:320 +0x33c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0004a28c0, {0x19bd558, 0xc000526f00})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:230 +0x325
```

```yaml
apiVersion: mongodbcommunity.mongodb.com/v1
kind: MongoDBCommunity
metadata:
  name: ...
spec:
  members: 1
  type: ReplicaSet
  version: "5.0.6"
  security:
    tls:
      enabled: true
      certificateKeySecretRef:
        name: mongodb-cert-secret
      #missing caConfigMapRef
```

--- a/controllers/mongodb_tls.go
+++ b/controllers/mongodb_tls.go
@@ -162,12 +162,17 @@ func getCaCrt(cmGetter configmap.Getter, secretGetter secret.Getter, mdb mdbv1.M
 	if mdb.Spec.Security.TLS.CaCertificateSecret != nil {
 		caResourceName = mdb.TLSCaCertificateSecretNamespacedName()
 		caData, err = secret.ReadStringData(secretGetter, caResourceName)
-	} else {
+	} else if mdb.Spec.Security.TLS.CaConfigMap != nil {
 		caResourceName = mdb.TLSConfigMapNamespacedName()
 		caData, err = configmap.ReadData(cmGetter, caResourceName)
 	}
+
 	if err != nil {
 		return "", err
+	}
+
+	if caData == nil {
+		return "", fmt.Errorf("TLS field requires a reference to the CA certificate which signed the server certificates. Neither secret (field caCertificateSecretRef) not configMap (field CaConfigMap) reference present")
 	}
 
 	if cert, ok := caData[tlsCACertName]; !ok || cert == "" {

--- a/controllers/mongodb_tls_test.go
+++ b/controllers/mongodb_tls_test.go
@@ -371,7 +371,7 @@ func TestTLSConfig_ReferencesToCACertAreValidated(t *testing.T) {
 
 			assert.NoError(t, err)
 
-			r := NewReconciler(kubeClient.NewManagerWithClient(cli))
+			r := NewReconciler(mgr)
 
 			_, err = r.validateTLSConfig(mdb)
 			if tc.expectedError != nil {

--- a/controllers/mongodb_tls_test.go
+++ b/controllers/mongodb_tls_test.go
@@ -2,6 +2,7 @@ package controllers
 
 import (
 	"context"
+	"errors"
 	"testing"
 
 	corev1 "k8s.io/api/core/v1"
@@ -332,66 +333,55 @@ func TestPemSupport(t *testing.T) {
 }
 
 func TestTLSConfig_ReferencesToCACertAreValidated(t *testing.T) {
-	t.Run("Success if reference to CA cert provided via secret", func(t *testing.T) {
-		mdb := newTestReplicaSetWithTLSCaCertificateReferences(&mdbv1.LocalObjectReference{
-			Name: "certificateKeySecret"}, nil)
+	type args struct {
+		caConfigMap         *mdbv1.LocalObjectReference
+		caCertificateSecret *mdbv1.LocalObjectReference
+		expectedError       error
+	}
+	tests := map[string]args{
+		"Success if reference to CA cert provided via secret": {
+			caConfigMap: &mdbv1.LocalObjectReference{
+				Name: "ccertificateKeySecret"},
+			caCertificateSecret: nil,
+		},
+		"Success if reference to CA cert provided via config map": {
+			caConfigMap: nil,
+			caCertificateSecret: &mdbv1.LocalObjectReference{
+				Name: "caConfigMap"},
+		},
+		"Succes if reference to CA cert provided both via secret and configMap": {
+			caConfigMap: &mdbv1.LocalObjectReference{
+				Name: "certificateKeySecret"},
+			caCertificateSecret: &mdbv1.LocalObjectReference{
+				Name: "caConfigMap"},
+		},
+		"Failure if reference to CA cert is missing": {
+			caConfigMap:         nil,
+			caCertificateSecret: nil,
+			expectedError:       errors.New("TLS field requires a reference to the CA certificate which signed the server certificates. Neither secret (field caCertificateSecretRef) not configMap (field CaConfigMap) reference present"),
+		},
+	}
+	for testName, tc := range tests {
+		t.Run(testName, func(t *testing.T) {
+			mdb := newTestReplicaSetWithTLSCaCertificateReferences(tc.caConfigMap, tc.caCertificateSecret)
 
-		mgr := kubeClient.NewManager(&mdb)
-		cli := mdbClient.NewClient(mgr.GetClient())
-		err := createTLSSecret(cli, mdb, "CERT", "KEY", "")
+			mgr := kubeClient.NewManager(&mdb)
+			cli := mdbClient.NewClient(mgr.GetClient())
+			err := createTLSSecret(cli, mdb, "cert", "key", "pem")
 
-		assert.NoError(t, err)
+			assert.NoError(t, err)
 
-		r := NewReconciler(mgr)
+			r := NewReconciler(mgr)
 
-		_, err = r.validateTLSConfig(mdb)
-		assert.Nil(t, err)
-	})
-	t.Run("Success if reference to CA cert provided via config map", func(t *testing.T) {
-		mdb := newTestReplicaSetWithTLSCaCertificateReferences(nil, &mdbv1.LocalObjectReference{
-			Name: "caConfigMap"})
+			_, err = r.validateTLSConfig(mdb)
+			if tc.expectedError != nil {
+				assert.EqualError(t, err, tc.expectedError.Error())
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
 
-		mgr := kubeClient.NewManager(&mdb)
-		cli := mdbClient.NewClient(mgr.GetClient())
-		err := createTLSSecret(cli, mdb, "CERT", "KEY", "")
-
-		assert.NoError(t, err)
-
-		r := NewReconciler(mgr)
-
-		_, err = r.validateTLSConfig(mdb)
-		assert.Nil(t, err)
-	})
-	t.Run("Succes if reference to CA cert provided both via secret and configMap", func(t *testing.T) {
-		mdb := newTestReplicaSetWithTLSCaCertificateReferences(&mdbv1.LocalObjectReference{
-			Name: "certificateKeySecret"}, &mdbv1.LocalObjectReference{
-			Name: "caConfigMap"})
-		mgr := kubeClient.NewManager(&mdb)
-		cli := mdbClient.NewClient(mgr.GetClient())
-		err := createTLSSecret(cli, mdb, "CERT", "KEY", "")
-
-		assert.NoError(t, err)
-
-		r := NewReconciler(mgr)
-
-		_, err = r.validateTLSConfig(mdb)
-		assert.Nil(t, err)
-	})
-	t.Run("Failure if reference to CA cert is missing", func(t *testing.T) {
-		mdb := newTestReplicaSetWithTLSCaCertificateReferences(nil, nil)
-
-		mgr := kubeClient.NewManager(&mdb)
-		cli := mdbClient.NewClient(mgr.GetClient())
-		err := createTLSSecret(cli, mdb, "CERT", "KEY", "")
-
-		assert.NoError(t, err)
-
-		r := NewReconciler(mgr)
-
-		_, err = r.validateTLSConfig(mdb)
-		assert.Error(t, err)
-		assert.Contains(t, err.Error(), "TLS field requires a reference to the CA certificate which signed the server certificates. Neither secret (field caCertificateSecretRef) not configMap (field CaConfigMap) reference present")
-	})
 }
 
 func createTLSConfigMap(c k8sClient.Client, mdb mdbv1.MongoDBCommunity) error {

--- a/controllers/mongodb_tls_test.go
+++ b/controllers/mongodb_tls_test.go
@@ -341,7 +341,7 @@ func TestTLSConfig_ReferencesToCACertAreValidated(t *testing.T) {
 	tests := map[string]args{
 		"Success if reference to CA cert provided via secret": {
 			caConfigMap: &mdbv1.LocalObjectReference{
-				Name: "ccertificateKeySecret"},
+				Name: "certificateKeySecret"},
 			caCertificateSecret: nil,
 		},
 		"Success if reference to CA cert provided via config map": {
@@ -371,7 +371,7 @@ func TestTLSConfig_ReferencesToCACertAreValidated(t *testing.T) {
 
 			assert.NoError(t, err)
 
-			r := NewReconciler(mgr)
+			r := NewReconciler(kubeClient.NewManagerWithClient(cli))
 
 			_, err = r.validateTLSConfig(mdb)
 			if tc.expectedError != nil {
@@ -401,7 +401,8 @@ func createTLSConfigMap(c k8sClient.Client, mdb mdbv1.MongoDBCommunity) error {
 func createTLSSecretWithNamespaceAndName(c k8sClient.Client, namespace string, name string, crt string, key string, pem string) error {
 	sBuilder := secret.Builder().
 		SetName(name).
-		SetNamespace(namespace)
+		SetNamespace(namespace).
+		SetField(tlsCACertName, "CERT")
 
 	if crt != "" {
 		sBuilder.SetField(tlsSecretCertName, crt)

--- a/controllers/mongodb_tls_test.go
+++ b/controllers/mongodb_tls_test.go
@@ -328,7 +328,69 @@ func TestPemSupport(t *testing.T) {
 		err = r.ensureTLSResources(mdb)
 		assert.Error(t, err)
 		assert.Contains(t, err.Error(), `if all of "tls.crt", "tls.key" and "tls.pem" are present in the secret, the entry for "tls.pem" must be equal to the concatenation of "tls.crt" with "tls.key"`)
+	})
+}
 
+func TestTLSConfig_ReferencesToCACertAreValidated(t *testing.T) {
+	t.Run("Success if reference to CA cert provided via secret", func(t *testing.T) {
+		mdb := newTestReplicaSetWithTLSCaCertificateReferences(&mdbv1.LocalObjectReference{
+			Name: "certificateKeySecret"}, nil)
+
+		mgr := kubeClient.NewManager(&mdb)
+		cli := mdbClient.NewClient(mgr.GetClient())
+		err := createTLSSecret(cli, mdb, "CERT", "KEY", "")
+
+		assert.NoError(t, err)
+
+		r := NewReconciler(mgr)
+
+		_, err = r.validateTLSConfig(mdb)
+		assert.Nil(t, err)
+	})
+	t.Run("Success if reference to CA cert provided via config map", func(t *testing.T) {
+		mdb := newTestReplicaSetWithTLSCaCertificateReferences(nil, &mdbv1.LocalObjectReference{
+			Name: "caConfigMap"})
+
+		mgr := kubeClient.NewManager(&mdb)
+		cli := mdbClient.NewClient(mgr.GetClient())
+		err := createTLSSecret(cli, mdb, "CERT", "KEY", "")
+
+		assert.NoError(t, err)
+
+		r := NewReconciler(mgr)
+
+		_, err = r.validateTLSConfig(mdb)
+		assert.Nil(t, err)
+	})
+	t.Run("Succes if reference to CA cert provided both via secret and configMap", func(t *testing.T) {
+		mdb := newTestReplicaSetWithTLSCaCertificateReferences(&mdbv1.LocalObjectReference{
+			Name: "certificateKeySecret"}, &mdbv1.LocalObjectReference{
+			Name: "caConfigMap"})
+		mgr := kubeClient.NewManager(&mdb)
+		cli := mdbClient.NewClient(mgr.GetClient())
+		err := createTLSSecret(cli, mdb, "CERT", "KEY", "")
+
+		assert.NoError(t, err)
+
+		r := NewReconciler(mgr)
+
+		_, err = r.validateTLSConfig(mdb)
+		assert.Nil(t, err)
+	})
+	t.Run("Failure if reference to CA cert is missing", func(t *testing.T) {
+		mdb := newTestReplicaSetWithTLSCaCertificateReferences(nil, nil)
+
+		mgr := kubeClient.NewManager(&mdb)
+		cli := mdbClient.NewClient(mgr.GetClient())
+		err := createTLSSecret(cli, mdb, "CERT", "KEY", "")
+
+		assert.NoError(t, err)
+
+		r := NewReconciler(mgr)
+
+		_, err = r.validateTLSConfig(mdb)
+		assert.Error(t, err)
+		assert.Contains(t, err.Error(), "TLS field requires a reference to the CA certificate which signed the server certificates. Neither secret (field caCertificateSecretRef) not configMap (field CaConfigMap) reference present")
 	})
 }
 

--- a/controllers/replica_set_controller.go
+++ b/controllers/replica_set_controller.go
@@ -73,7 +73,6 @@ func NewReconciler(mgr manager.Manager) *ReplicaSetReconciler {
 	mgrClient := mgr.GetClient()
 	secretWatcher := watch.New()
 	configMapWatcher := watch.New()
-
 	return &ReplicaSetReconciler{
 		client:           kubernetesClient.NewClient(mgrClient),
 		scheme:           mgr.GetScheme(),

--- a/controllers/replicaset_controller_test.go
+++ b/controllers/replicaset_controller_test.go
@@ -87,6 +87,15 @@ func newScramReplicaSet(users ...mdbv1.MongoDBUser) mdbv1.MongoDBCommunity {
 }
 
 func newTestReplicaSetWithTLS() mdbv1.MongoDBCommunity {
+	return newTestReplicaSetWithTLSCaCertificateReferences(&mdbv1.LocalObjectReference{
+		Name: "caConfigMap",
+	},
+		&mdbv1.LocalObjectReference{
+			Name: "certificateKeySecret",
+		})
+}
+
+func newTestReplicaSetWithTLSCaCertificateReferences(caConfigMap, caCertificateSecret *mdbv1.LocalObjectReference) mdbv1.MongoDBCommunity {
 	return mdbv1.MongoDBCommunity{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:        "my-rs",
@@ -101,10 +110,9 @@ func newTestReplicaSetWithTLS() mdbv1.MongoDBCommunity {
 					Modes: []mdbv1.AuthMode{"SCRAM"},
 				},
 				TLS: mdbv1.TLS{
-					Enabled: true,
-					CaConfigMap: &mdbv1.LocalObjectReference{
-						Name: "caConfigMap",
-					},
+					Enabled:             true,
+					CaConfigMap:         caConfigMap,
+					CaCertificateSecret: caCertificateSecret,
 					CertificateKeySecret: mdbv1.LocalObjectReference{
 						Name: "certificateKeySecret",
 					},

